### DISCUSSION TITLE
Async calls to private field properties

### DISCUSSION THREAD

**Discussion by JacoJordaan** [MarimerLLC/csla#1636] [General]
I have a use case that I am not sure what is the best approach to handle when changing my objects to use async for Blazor wasm compatibility.

Currently I have many using reference properties in my business objects. 
The main reasoning is that these are used inside the business object, but must not be serialized as part of the object and also just loaded when demanded.
Also, some of these private field properties are cached - loaded once when required and then loaded from memory. 
Many of these properties are used inside business rules, so when a business rule run and there is a condition required for using the property, the business object is loaded into property and used appropriately. 

For Blazor wasm I need to change property as follows:
From:
```
        public static readonly PropertyInfo<BrandList> BrandListProperty = RegisterProperty<BrandList>(c => c.BrandList, RelationshipTypes.PrivateField);

        [NotUndoable, NonSerialized]
        private BrandList _brandList = null;

        public BrandList BrandList
        {
            get
            {
                if (_brandList == null)
                {
                    _brandList = BrandList.GetBrandList();
                }
                return GetProperty(BrandListProperty, _brandList);
            }
        }

```
To:
```
       public static readonly PropertyInfo<BrandList> BrandListProperty = RegisterProperty<BrandList>(c => c.BrandList, RelationshipTypes.PrivateField);

        [NotUndoable, NonSerialized]
        private BrandList _brandList = null;

        public BrandList BrandList
        {
            get
            {
                if (_brandList == null)
                {
                    Task.Run(async () =>
                      {
                          _brandList = await BrandList.GetBrandListAsync();
                          OnPropertyChanged(nameof(BrandList));
                      });
                }
                return GetProperty(BrandListProperty, _brandList);
            }
        }
```

To problem now is that when my business rule is trying to load the property for the first time, it is not yet available and my rule fails.

**What is the best approach in these situations?**

I was thinking of changing the rule to not run when the property is still null. Add the same rule to the BrandList as primary property and re-run the rule when the BrandList rule changes. I think this may work, but it is not ideal. Any other ideas would be appreciated.

**Comment by ajj7060** [comment]
Use LazyGetPropertyAsync instead of GetProperty.  That will also allow you to get rid of the private backing field completely, which are no longer supported in BOs.

**Reply by JacoJordaan** [reply]
Thanks for the feedback.
I have looked at LazyGetPropertyAsync, but this also has the same issue - the property is null when while loading the property causing the business rule to fail.

I have managed to change many of the properties to load before the business object loads by using a unit of work object.

**Reply by ajj7060** [reply]
Right, the value will initially be default until the get property finishes.  I see your GetBrandList doesn't take properties; what if you made your rule async and have it do the GetBrandList instead?

That's typically what I would do; actually if instead of calling the BrandList factory method directly, you could create an IBrandListFactory interface, and create two classes; one that just calls the static factory method for BrandList and the other that will check if the list is already cached, and if not will fetch & cache the result.  

The benefit to this approach is that your caching can be separated (and possibly using something like Microsoft.Caching), and your BO won't need to have a Brands property at all just to support the caching.
